spdmlib/src/crypto/x509v3.rs

```diff
@@ -257,6 +293,121 @@ fn check_public_key_info(data: &[u8]) -> SpdmResult<usize> {
     check_and_skip_common_sequence(data)
 }
 
+fn check_and_get_extensions(data: &[u8]) -> SpdmResult<(usize, &[u8])> {
+    let len = data.len();
+    if len < 1 || data[0] != ASN1_TAG_EXPLICIT_EXTENSION {
+        Ok((len, &data[0..]))
+    } else {
+        let (payload_length, bytes_consumed) = check_length(&data[1..])?;
+        if len < 1 + bytes_consumed + payload_length {
+            Err(SPDM_STATUS_VERIF_FAIL)
+        } else {
+            Ok((
+                1 + bytes_consumed + payload_length,
+                &data[1 + bytes_consumed..1 + bytes_consumed + payload_length],
+            ))
+        }
+    }
+}
+
+fn get_key_usage_value(data: &[u8]) -> SpdmResult<(bool, u8)> {
+    let len = data.len();
+    let mut key_usage_value_len = 0;
+    let mut key_usage_value_index: usize = 0;
+    if len < 3 {
+        Ok((false, 0x00))
+    } else {
+        let mut find_key_usage: bool = false;
+        for index in 0..len - 3 {
+            if object_identifiers_are_same(&data[index..index + 3], OID_KEY_USAGE) {
+                let (key_usage_bytes_consumed, key_usage) =
+                    check_and_get_common_tag(&data[index + 3..])?;
+                if key_usage[0] != ASN1_TAG_BIT_STRING {
+                    find_key_usage = false;
+                } else {
+                    find_key_usage = true;
+                    (key_usage_value_len, _) = check_length(&key_usage[1..])?;
+                    key_usage_value_index =
+                        index + 3 + key_usage_bytes_consumed - key_usage_value_len;
+                }
+                break;
+            } else {
+                continue;
+            }
+        }
+        if find_key_usage {
+            Ok((true, data[key_usage_value_index + key_usage_value_len - 1]))
+        } else {
+            Ok((false, 0x00))
+        }
+    }
+}
+
+fn check_extensions_spdm_oid(data: &[u8], is_leaf_cert: bool) -> bool {
+    let len = data.len();
+    let oid_len = 10;
+    let mut spdm_extension_oid_find_success: bool = false;
+    let mut mutable_cert_oid_find_success: bool = false;
+    let mut hardware_identity_oid_find_success: bool = false;
+    let mut responder_auth_oid_find_success: bool = false;
+    let mut requester_auth_oid_find_success: bool = false;
+    if len < 2 + oid_len {
+        info!("length of extensions is too small, no spdm oids\n");
+        true
+    } else {
+        let extensions_end_index_oid = len - oid_len;
+        for index in 2..extensions_end_index_oid {
+            if object_identifiers_are_same(&data[index..index + oid_len], OID_DMTF_SPDM_DEVICE_INFO)
+                && (data[index - 2] == ASN1_TAG_OID)
```

Comment:
1) It is quite unusual that use negative value as index -  `[index - 2]`. 
2) It redundant to have data[index - 2] check in each `if`.
3) It is not efficient to move index by 1 in `for`.

Reply:
Thanks! I have changed the check_extensions_spdm_oid logic:
-Get extensions data from cert first
---Get the extnID of the first extension sequence, if not the target, then jump to next extension sequence
-----Search the target id-DMTF-spdm belong to this extension sequence
When didn't find the extension identify under the extension sequence or some length check fail, it will return error

RFC 5280 Page 17 extnID under extension sequence:
<img width="371" alt="image" src="https://github.com/intel/rust-spdm/assets/109128373/683a2b0e-7742-44e4-8d02-96a66317945c">

The id-DMTF-spdm should under the following extension sequence base on the SPDM spec, and I follow this to search the target id-DMTF-spdm OID to ensure they are in the right place:
<img width="250" alt="image" src="https://github.com/intel/rust-spdm/assets/109128373/7a5997ed-77dc-4dcc-89e1-78473c13e5aa">